### PR TITLE
Hacktoberfest Re-organisation

### LINKED ISSUE #861: `.timeleft` command should be a subcommand of hacktoberfest
## Description
Currently we have a `.timeleft` command under the hacktoberfest cog. However, it's a top-level command.
I'd like to move it as a subcommand so you would have to do `.hacktoberfest timeleft`.

## Reasoning
Hacktoberfest is no longer going to be a major event we participate in and we have other events that I think would be better suited to this type of countdown. Advent of Code already has a `countdown` subcommand and that is a much bigger event for our server.

## Proposed Implementation
It should be moved under the hacktoberfest cog and be a subcommand.
This may also require a general re-work of this cog so it's more cohesive.

- [ ] I'd like to implement this feature myself
- [X] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by TizzySaurus**
## Relevant Issues
Closes #861.

## Description

- Moved all hacktoberfest commands into `_cog.py`
- Moved all hacktoberfest util functions into `_utils.py`
- Deleted the three hacktoberfest command files (since merged into `_cog.py` and `_utils`)
- Migrated all hacktoberfest commands to be under the `Hacktoberfest` cog and `.hacktoberfest` command (alias `.hacktober`)
- Renamed some commands (`.hackstats` --> `.hacktoberfest stats`, `.hacktoberissues` --> `.hacktoberfest issues` and `.timeleft` --> `.hacktoberfest timeleft`)
- Added a temp-fix to the already broken `.hacktstats` courtesy of @hedyhli. I believe they're planning to write a more permanent fix soon:tm:.

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

**Comment by TizzySaurus** [comment]
> Hey @TizzySaurus (for your PR description) the tempfix was for `.hackstats` not `.hacktoberissues` :)

Fixed, thanks :+1:

**Comment by TizzySaurus** [comment]
> @TizzySaurus is this PR only focused on migrating the code as-is and the restructuring or are you wanting to refactor the three commands as well? Many of Jason's suggestions are on (mostly 😃) copied original implementations (from #7, #46, #296, #466, #475, [047902d](https://github.com/python-discord/sir-lancebot/commit/047902d599bb87b7227a259fe63fa3a3f34b1d71), possibly more). IMO it should just copy the functions over for now plus many adjustments (which you've done) and then refactor them later in separate PRs (someone else could do this) since this is only about putting the three subcommands under a single `hacktoberfest` command.

I may redo certain easy parts, but I won't be completely refactoring the code. I agree with you that refactoring is outside the scope of this PR.

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:73]: Have done this locally :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:447]: Have done this locally :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:449]: I guess the idea is that the footer would be more obvious the link is there and then the `.url` is what you can click to open it? I don't really know, but think for now I'll leave this.

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:96]: I believe this was deliberately done as `INFO` so that it would show on our logs online (which doesn't show `TRACE` level).

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:105]: Yeah, I'm inclined to agree with this. It means we don't have this specific error handling but we have generic handling for when there's a missing argument to a command.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:108]: Done locally, with a minor improvement to the if statements.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:184]: Done locally, with comments explaining each line :+1:

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:34]: I prefer the latter, so have done that locally :+1:

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:30]: Yep, have removed the `__init__` locally :+1:

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:102]: I'm not too sure but I'm going to leave this one as-is. Perhaps @hedyhli can look into this for one of their Hacktoberfest PRs.

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:70]: Have done this locally :+1:

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:95]: Have done this locally :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:446]: Parsing the markdown seems too big of a thing for the scope of this PR. Perhaps @hedyhli would like to do this as one of their hacktoberfest PRs.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:43]: Have done this locally :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:30]: I think I'll leave this now as I don't really have time to delve into this. Perhaps you can look into this for one of your hacktoberfest PRs.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [reply bot/exts/events/hacktoberfest/_utils.py:446]: Shouldn't the markdown already render like the GitHub webhook embeds?

**Comment by TizzySaurus** [comment]
@Shivansh-007 @hedyhli Have just pushed [#fca5a45](https://github.com/python-discord/sir-lancebot/pull/919/commits/fca5a4598eea6e44fe5d39540ff808e069b59f82) which fixes all the reviews I said I'd do :+1:

File: bot/exts/events/hacktoberfest/_cog.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_cog.py:41]: I generally like to recommend aliases in the help or suggestion text
```suggestion
        If the command is run with beginner (`.hacktober issue beginner`):
```

File: bot/exts/events/hacktoberfest/_cog.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_cog.py:77]:
```suggestion
            log.info(f"Getting stats for '{github_username}' as requested by {ctx.author.id}")
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_utils.py:50]:
```suggestion
    "https://api.github.com/search/issues?"  # base url
    "per_page=100"                           # limit results per-page returned by API to 100 (the maximum)
    "&q="                                    # add query parameters
    "is:issue+"                              # is an issue
    "state:open+"                            # that's open
    "label:hacktoberfest+"                   # has `hacktoberfest` label
    "language:python"                        # in Python
```
Not sure if the width would pass lint test

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_utils.py:182]:
```suggestion
        f"+type:pr"  # Only get PR if it's:
        f"+is:public"  # public
        f"+author:{quote_plus(github_username)}"  # made by the user's github username
        f"+-is:draft"   # not a draft
        f"+created:{CURRENT_YEAR}-09-30T10:00Z..{CURRENT_YEAR}-11-01T12:00Z"  # made in October
        f"&per_page=100"  # Limit results per-page returned from API (100 is the maximum)
```
The `...` looks a bit redundant, up to you though.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_utils.py:333]:
```suggestion
    'hacktoberfest-accepted'.
```

**Review by hedyhli — CHANGES_REQUESTED** [review]
Almost there

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [reply bot/exts/events/hacktoberfest/_utils.py:30]: Sure thing

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:41]: I'd rather stick to the full command myself. It just feels more "complete".

File: bot/exts/events/hacktoberfest/_cog.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_cog.py:77]: Good catch :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:333]: Good catch :+1:

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:50]: The width does pass and I like it, so have done this :+1:

I'll also remove the `...` since I think it adds more noise rather than helping with flow as intended.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (TizzySaurus) [reply bot/exts/events/hacktoberfest/_utils.py:182]: Agree, have removed. Also aligned the comments in a way that looks neater.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_utils.py:147]:
```suggestion
    logging.info(f"Hacktoberfest embed built for GitHub user '{github_username}'")
```
So it looks similar to line 111

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_utils.py:296]:
```suggestion
        # If API request is unsuccessful it will be a dict with the error in 'message'
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [reply bot/exts/events/hacktoberfest/_utils.py:446]: Yeah, most of the markdown is passed but HTML tags like `p`, `details` aren't used so often, so this feature is not really needed, but it would be good to have it.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:419]:
```suggestion
            return
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:397]: Not really helpful, could you add a bit more information on this logging statement. And we can bump such statements to `info` IMO.

Secondly, I don't think `debug` level logging statements are visible in the pydis production environment (Confirmed by joe).

File: bot/exts/events/hacktoberfest/_cog.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_cog.py:23]: Could you add type hints for this variable?

File: bot/exts/events/hacktoberfest/_cog.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_cog.py:25]: Could you add type hints for this variable?

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:310]:
```suggestion
    return any(item.get("status") == "APPROVED" for item in jsonresp2)
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:272]:
```suggestion
    return any(
        any(label["name"].casefold() == item for label in pr["labels"])
        for item in labels
    )
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:225]: You can merge the nested if conditions

```suggestion
        if has_label(item, ["invalid", "spam"]) and not await is_accepted(
            bot, itemdict
        ):
            continue
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:120]:
```suggestion
    n = len(accepted) + len(in_review)  # Total number of PRs
    remaining_prs = PRS_FOR_SHIRT - n   # Remaining PRs for completing hacktoberfest
    if n >= PRS_FOR_SHIRT:
        shirtstr = f"**{github_username} is eligible for hacktoberfest swag!**"
    else:
        shirtstr = f"**{github_username} is {remaining_prs} PR{'s' if remaining_prs > 0 else ''} away from being eligible for hacktoberfest swag!**"
```

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:124]: This could be made a constant as the hacktoberfest colours are changing every year... or even smarter at the start of hacktoberfest, you can just fetch the GitHub profile picture and extract the most dominant colour.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:106]: Could be merged into a single if...else conditional.

File: bot/exts/events/hacktoberfest/_utils.py
Comment (Shivansh-007) [inline bot/exts/events/hacktoberfest/_utils.py:43]: Could be renamed to `ISSUES_REQUEST_URL`, the name `URL` isn't really helpful.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [reply bot/exts/events/hacktoberfest/_utils.py:43]: Suggested [here](https://github.com/python-discord/sir-lancebot/pull/919#discussion_r732366679), @TizzySaurus you forgot to update the variable name 😄

**Review by hedyhli — CHANGES_REQUESTED** [review]
Just some trivial things apart from your previous commit breaking `hacktober stats` (at least for me)

File: bot/exts/events/hacktoberfest/_utils.py
Comment (hedyhli) [reply bot/exts/events/hacktoberfest/_utils.py:446]: It would require flattening out (transforming HTML to plaintext) but I don't really think it's that much of a problem. Users should go to the GitHub issue page anyway to see other information like assignees and linked PRs. If you'd still like that you should open a issue and we can discuss about whether it should be added.

File: bot/exts/events/hacktoberfest/_cog.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_cog.py:35]:
```suggestion
    @in_month(Month.SEPTEMBER, Month.OCTOBER)
```
Perhaps people would like to start looking for issues beforehand if maintainers are already preparing for it

File: bot/exts/events/hacktoberfest/_cog.py
Comment (hedyhli) [inline bot/exts/events/hacktoberfest/_cog.py:53]:
```suggestion
    @hacktoberfest.group(invoke_without_command=True, root_aliases=('hackstats',))
```
How about some backwards compatibilily? :P
I think `hackstats` is unique enough unless we roll out some sort of hacking-related event/competition in the future

**Review by hedyhli — CHANGES_REQUESTED** [review]
Just two additional things I forgot in my last review (so we can get this merged quicker to hopefully sort out the other hacktober-related things before the end of november 😃)

**Review by Bluenix2 — COMMENTED** [review]
We're loosing all the git blame history here.. can you try using [`git mv`](https://git-scm.com/docs/git-mv)?

That may save *something*?

**Review by Bluenix2 — CHANGES_REQUESTED** [review]
Actually I will Request changes so that this can't be merged until we have looked into this more.

**Comment by lemonsaurus** [comment]
mm, @Bluenix2, you make a fair point, but we shouldn't block this merge over this small amount of git blame. It's fine to lose a little git blame in a refactor like this, and cleaning this up now after the fact is going to be quite annoying. I don't want this PR getting closed or abandoned over a technicality that might never have any real practical use for us, especially after all these wonderful people spent time reviewing it.

**Comment by hedyhli** [comment]
Hey @TizzySaurus, would you be able to continue working on this?

**Comment by Xithrius** [comment]
I'll be closing this PR due to staleness.


@hedyhli if you'd like to work on this, please ping me and I'll assign you, and the PR can be re-opened.

**Pull request closed without merging**
